### PR TITLE
Add tus for large file uploads 

### LINKED ISSUE #2: Add tus & JWT support as stop-gap before isomorphic skynet-js
We're starting to see more server-side use-cases for Skynet. I'd like to propose we add `tusUpload` along with a simple way to add your account JWT.

I think all other actions are easy to do by just using both libraries, but until we have a plan on our isomorphic library, this will fill the gaps. I'd also suggest moving it to `@skynetlabs` org on npm, breaking the nebulous branding.

The `tus-js-client` is already isomorphic, so I think this is low-hanging fruit compared to developers trying to figure out tus implementations on their own.

### PR DESCRIPTION
**Pull request by m-cat**
Closes #2

### CODE REVIEW

**Review by dghelm — APPROVED** [review]

File: src/utils.js [hunk 1]
```diff
@@ -31,6 +33,24 @@ function defaultPortalUrl() {
...
+function getFileMimeType(filename) {
+  let { ext } = path.parse(filename);
+  ext = trimPrefix(ext, ".");
+  if (ext !== "") {
+    const mimeType = mime.getType(ext);
+    if (mimeType) {
+      return mimeType;
+    }
+  }
+  return "";
```
Comment (ro-tex) [inline src/utils.js:51]: Should we default to an empty string or `application/octet-stream`?

**Review by ro-tex — APPROVED** [review]

File: src/client.js [hunk 2]
```diff
@@ -57,7 +59,29 @@ class SkynetClient {
...
+  }
+  // Set some headers from common options.
+  if (customUserAgent) {
+    headers["User-Agent"] = customUserAgent;
+  }
+  if (customCookie) {
+    headers["Cookie"] = customCookie;
+  }
+  return headers;
+}
```
Comment (kwypchlo) [inline src/client.js:82]: you're mutating `headers` if they are passed as an argument - you should assign to new object

File: src/upload.js [hunk 3]
```diff
@@ -4,46 +4,115 @@ const FormData = require("form-data");
...
 };
 
-SkynetClient.prototype.uploadFile = function (path, customOptions = {}) {
+SkynetClient.prototype.uploadFile = async function (path, customOptions = {}) {
   const opts = { ...defaultUploadOptions, ...this.customOptions, ...customOptions };
 
-  const formData = new FormData();
-  const filename = opts.customFilename ? opts.customFilename : "";
-  formData.append(opts.portalFileFieldname, fs.createReadStream(path), filename);
+  const sizeInBytes = fs.statSync(path).size;
```
Comment (kwypchlo) [inline src/upload.js:39]: use async stat, it's already na async function anyway https://nodejs.org/api/fs.html#fs_fspromises_stat_path_options

File: src/utils.js (hunk 1 above)
Comment (kwypchlo) [reply src/utils.js:51]: There was a discussion about that here https://stackoverflow.com/questions/1176022/unknown-file-type-mime

Most upvoted answer is "You can use application/octet-stream for unknown types." but the other one says:

> RFC resources:
We should use RFC-7231 (HTTP/1.1 Semantics and Content) as reference instead of RFC-2046 (Media Types) because question was clearly about HTTP Content-Type.
>
> Also RFC-2046 does not clearly define unknown types but RFC-7231 does.
>
> Short answer:
Do not send MIME type for unknown data.
To be more clear: Do not use Content-Type header at all.

I would side with not setting the mime type at all for unknown type.

File: src/utils.js [hunk 4]
```diff
@@ -42,6 +62,27 @@ function makeUrl() {
...
+      break;
+    }
+    str = str.slice(prefix.length);
+    if (limit) {
+      limit -= 1;
+    }
+  }
+  return str;
+}
+
```
Comment (kwypchlo) [inline src/utils.js:85]: I think this might look a little more readable

```suggestion
function trimPrefix(str, prefix, limit) {
  const matcher = new RegExp(`^(${prefix}){0,${limit ?? ""}}`);
  return str.replace(matcher, "")
}

```

**Review by kwypchlo — CHANGES_REQUESTED** [review]

File: scripts/upload.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+const process = require("process");
+
+const { SkynetClient } = require("..");
+
+const client = new SkynetClient();
+
+const promises = process.argv.slice(2).map((path) => client.uploadFile(path));
+(async () => {
+  const skylinks = await Promise.all(promises);
```
Comment (peterjan) [inline scripts/upload.js:13]: `Promise.all` rejects immediately. I think it would be better to use `Promise.allSettled`.

File: src/client.js [hunk 6]
```diff
@@ -27,21 +28,22 @@ class SkynetClient {
...
    */
   executeRequest(config) {
-    const url = makeUrl(config.portalUrl, config.endpointPath, config.extraPath ? config.extraPath : "");
-
-    if (!config.headers) {
-      config.headers = {};
-    }
-    if (config.customUserAgent) {
-      config.headers["User-Agent"] = config.customUserAgent;
+    let url;
```
Comment (peterjan) [inline src/client.js:31]:
```
let url = config.url;
if (!url) {
    url = makeUrl(config.portalUrl, config.endpointPath, config.extraPath ? config.extraPath : "");
}
```

File: src/upload.js [hunk 7]
```diff
@@ -4,46 +4,115 @@ const FormData = require("form-data");
...
+  const filename = opts.customFilename ? opts.customFilename : p.basename(path);
 
+  if (sizeInBytes < opts.largeFileSize) {
+    return await uploadSmallFile(this, path, filename, opts);
+  } else {
```
Comment (peterjan) [inline src/upload.js:44]: redundant `else`

File: src/upload.js [hunk 8]
```diff
@@ -4,46 +4,115 @@ const FormData = require("form-data");
...
+    return await uploadLargeFile(this, path, filename, sizeInBytes, opts);
+  }
+};
+
+async function uploadSmallFile(client, path, filename, opts) {
   const params = {};
   if (opts.dryRun) params.dryrun = true;
 
+  const formData = new FormData();
+  formData.append(opts.portalFileFieldname, fs.createReadStream(path), filename);
```
Comment (peterjan) [inline src/upload.js:54]: should we validate the `portalFileFieldname` is not empty?

File: src/utils.js [hunk 9]
```diff
@@ -31,6 +33,24 @@ function defaultPortalUrl() {
...
 }
 
+/**
+ * Get the file mime type. Try to guess the file type based on the extension.
+ *
+ * @param filename - The filename.
+ * @returns - The mime type.
+ */
+function getFileMimeType(filename) {
+  let { ext } = path.parse(filename);
```
Comment (peterjan) [inline src/utils.js:43]: `path.extname` probably better

File: scripts/upload.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Example script that uploads all paths passed in as CLI arguments.
+ */
+
+const process = require("process");
+
+const { SkynetClient } = require("..");
```
[the remark this answers is not in the document]
Comment (m-cat) [reply scripts/upload.js:7]: It's just a shorthand, e.g. `require("src")` is the same as `require("src/index.js")` by default. I can see how that can be confusing but I believe it's standard Javascript. I can change it if I'm mistaken about that.

File: src/utils.js (hunk 4 above)
Comment (m-cat) [reply src/utils.js:85]: I'm not too familiar with regexps and try not to use them. They tend to be hard to read and inefficient. I'm not sure that one will work if you substitute in `${prefix}` like that, and the second half with `limit` I unfortunately don't understand at all.

File: src/upload.js (hunk 7 above)
Comment (m-cat) [reply src/upload.js:44]: I think the version with `else` is more clear but I don't care either way.

File: src/upload.js (hunk 8 above)
Comment (m-cat) [reply src/upload.js:54]: We don't really do any validation of inputs in this SDK. I don't think it's worth hardening this one like we did with `skynet-js` as this will be archived once we have an isomorphic SDK.

File: src/utils.js (hunk 1 above)
Comment (ro-tex) [reply src/utils.js:51]: Works for me. :)

File: src/utils.js (hunk 4 above)
Comment (ro-tex) [reply src/utils.js:85]: I would also prefer the non-regexp version here. The regexp is shorter but it's a little harder to read and more prone to errors when modified. The while version is dead simple and immediately understandable.

**Review by ro-tex — APPROVED** [review]

File: src/utils.js (hunk 1 above)
Comment (peterjan) [reply src/utils.js:51]: I think what Karol means is that if `getFileMimeType(filename)` is an empty string, we should simply not pass it along with the metadata under the `filetype` property. Providing that's an option of course...

so:
```
const tusOpts = {
      ....
      metadata: {
        filename,
        filetype: getFileMimeType(filename),
      },
      ...
}
```
simply becomes:
```
const filetype = getFileMimeType(filename)
const metadata = filetype ? { filename, filetype } : { filename }
const tusOpts = {
      ....
      metadata,
      ...
}
```

File: src/utils.js (hunk 4 above)
Comment (peterjan) [reply src/utils.js:85]: The original `trimPrefix` implementation is already flawed because it handles `null` and `undefined` differently. It's because you strict compare it to `undefined`, which is always a bit dangerous. 

The `RegExp` only looks complicated but it's honestly just a group and a quantifier. I do think the bullish coalescing here makes it hard to read. Extracting it into a variable and adding a comment would make it easy to read imo.

```
const repeats = `{0,${limit ?? ""}}` // trim all if limit is null or undefined
const matcher = new RegExp(`^(${prefix})${repeats}`);
return str.replace(matcher, "")
```

File: src/upload.js (hunk 8 above)
Comment (peterjan) [reply src/upload.js:54]: not sure if that's a good enough reason to not do basic validation on inputs

Comment (peterjan) [reply src/upload.js:54]: In this particular case though I guess it's fine to leave it as is.

**Review by peterjan — CHANGES_REQUESTED** [review]
Thanks for addressing the review comments. Request changes only because the curr. implementation of `trimPrefix` is flawed, looks good otherwise.

File: src/utils.js (hunk 4 above)
Comment (peterjan) [reply src/utils.js:85]: Probably best to add a test case that covers both `null` and `undefined`.

Comment (m-cat) [reply src/utils.js:85]: The function is taken from `skynet-js` where we have Typescript typing. The type of `limit` is `limit?: number`, so if someone were to try to pass `null` it wouldn't compile.

I don't think a regexp here is appropriate and I'm not sure it's performant to be compiling it every time the function is called. Let's not spend more resources on a regexp solution here, let's just keep what is already there and known to work.

**Review by kwypchlo — APPROVED** [review]

File: src/utils.js (hunk 4 above)
Comment (peterjan) [reply src/utils.js]: How can we approve this PR though. It adds a function which we know is flawed. If you pass `null` or `undefined` you won't get the same outcome. If you don't want to adopt the regex, why wouldn't we update the code to work properly.

Comment (m-cat) [reply src/utils.js]: I added validation.

**Review by peterjan — APPROVED** [review]
